### PR TITLE
Introduce SystemInformation service, hook into UDC and server.

### PR DESCRIPTION
**Pull request by jakewins**
UDC has a complex dependency hierarchy, and because of that it uses reflection
to extract information from system components. This introduces a "SystemInformation"
service where components publish metadata and UDC can access it without reflection.

Also minor improvements to UDC resilience - only store up to 10 latest client
names and urlencode query parameters in UDC ping payload.

### CODE REVIEW

**Comment by jakewins** [comment]
This cleans some stuff up, but will mainly serve to allow NDP to publish Client name information to UDC together with the REST API, and hopefully it'll also make future modifications to UDC metadata easier..

**Comment by cleishm** [comment]
If we make UDC a concrete dependency of the kernel, would that simplify things?

**Comment by jakewins** [comment]
In the longer term, yes. However, it's a documented feature that you can remove the jar to disable UDC, so if we wanted to clean that up to just use the config toggle, we'd need to run a deprecation cycle on it like any other feature.

http://neo4j.com/docs/stable/usage-data-collector.html

So - I think we should do both, go through the motions of deprecating the remove-the-jar feature, move this in and then in 3.x we merge this service and UDC into a common thing. Sound ok?

**Comment by chrisvest** [comment]
@jakewins You have a compilation failure in CommunityNeoServer.

File: community/primitive-collections/src/main/java/org/neo4j/concurrent/RecentK.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    private final ConcurrentHashMap<Type, Slot> index = new ConcurrentHashMap<>();
+
+    /** Most recently seen item */
+    // guarded by synchronized
+    private Slot head;
+
+    /** Least recently used item */
+    // guarded by synchronized
+    private Slot tail;
```
Comment (chrisvest) [inline community/primitive-collections/src/main/java/org/neo4j/concurrent/RecentK.java:47]: The mechanism of locking is less important than the object that is the lock. I'd write "guarded by 'this'" or "guarded by synchronized(this)".

File: community/server/src/main/java/org/neo4j/server/modules/DBMSModule.java [hunk 2]
```diff
@@ -21,8 +21,10 @@
 
 import java.util.List;
 
+import org.neo4j.server.rest.dbms.AuthorizationFilter;
 import org.neo4j.server.rest.dbms.UserService;
 import org.neo4j.server.rest.discovery.DiscoveryService;
+import org.neo4j.server.rest.web.CollectUserAgentFilter;
 import org.neo4j.server.web.WebServer;
```
Comment (chrisvest) [inline community/server/src/main/java/org/neo4j/server/modules/DBMSModule.java:26]: These imports aren't used, are they?

**Comment by chrisvest** [comment]
@jakewins I think this is good to merge once you get it to compile, and the small comments are addressed.

**Comment by jakewins** [comment]
Modified it to work as a first step for moving UDC to kernel. Moving UDC turned out to be more complex, since it hooks into build infrastructure to track how artifacts have been produced and distributed. Leaving that for a follow-up commit.

**Pull request merged**